### PR TITLE
Adding Configuration Checker

### PR DESCRIPTION
**Pull request by kasundharmadasa**
## Purpose
 In WSO2 products there are configurations that are common to all products and specific to a product.  These configurations needs to be validated against a recommended set of configurations

## Goals
This tool can be used to validate the configurations.

## Approach
Therefore this tool uses a <application_directory>/Common/properties/parent.xml file to hold the common configuration files and a <application_directory>/Product/properties/child.xml to hold product specific configuration files.

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: external/config-checker/src/main/java/org/wso2/security/tools/conf/ConfigCheckerSettingsBuilder.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * Reading the config checker settings from a given config file.
+     *
+     * @param configFile
+     * @return
+     * @throws ConfigCheckerException
+     */
+    public static ConfigCheckerSettings getConfiguration(File configFile) throws ConfigCheckerException {
+        ConfigCheckerSettings configuration = null;
+
```
Comment (ayomawdb) [inline external/config-checker/src/main/java/org/wso2/security/tools/conf/ConfigCheckerSettingsBuilder.java]:
```suggestion
```

File: external/config-checker/src/main/java/org/wso2/security/tools/conf/ConfigCheckerSettingsBuilder.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+     * Reading the config checker settings from config-checker-settings.yaml.
+     *
+     * @return
+     * @throws ConfigCheckerException
+     */
+    public static ConfigCheckerSettings getConfiguration() throws ConfigCheckerException {
+
+        ConfigCheckerSettings configCheckerSettings = null;
+        configCheckerSettings = getConfiguration(new File(CONFIG_CHECKER_SETTINGS_FILE));
+
```
Comment (ayomawdb) [inline external/config-checker/src/main/java/org/wso2/security/tools/conf/ConfigCheckerSettingsBuilder.java]:
```suggestion
```

File: external/config-checker/src/main/java/org/wso2/security/tools/conf/ConfigCheckerSettingsBuilder.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    private static final String CONFIG_CHECKER_SETTINGS_FILE = "config-checker-settings.yaml";
+
+    /**
+     * Reading the config checker settings from config-checker-settings.yaml.
+     *
+     * @return
+     * @throws ConfigCheckerException
+     */
+    public static ConfigCheckerSettings getConfiguration() throws ConfigCheckerException {
+
```
Comment (ayomawdb) [inline external/config-checker/src/main/java/org/wso2/security/tools/conf/ConfigCheckerSettingsBuilder.java]:
```suggestion
```

**Comment by ayomawdb** [comment]
We need to move this project to `internal` folder.

**Comment by ayomawdb** [comment]
Also the main package used should be changed from `org.wso2.security.tools` to `org.wso2.security.tools.configchecker`.

File: external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java [hunk 4]
```diff
@@ -0,0 +1,601 @@
...
+                        }
+                    });
+        } catch (IOException e) {
+            throw new ConfigCheckerException("IO Exception occurred while retrieving the configuration files", e);
+        }
+        return paths;
+    }
+
+    /**
+     * This method is used to get the configuration file name.
```
Comment (ayomawdb) [inline external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java]: This description and the method name are contradicting. Shall we improve the comment?

File: external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java [hunk 5]
```diff
@@ -0,0 +1,601 @@
...
+    }
+
+    /**
+     * This method is used to get the configuration id from the <config> element.
+     *
+     * @param configNode
+     * @return configuration id
+     * @throws ConfigCheckerException
+     */
+    private static String getConfigID(Node configNode) throws ConfigCheckerException {
```
Comment (ayomawdb) [inline external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java]:
```suggestion
    private static String getConfigId(Node configNode) throws ConfigCheckerException {
```

You'll ave to correct usages of the function.

File: external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+                            configProperties.getLength(); configPropertiesIndex++) {
+
+                        if (configProperties.item(configPropertiesIndex).getNodeType() ==
+                                Node.CDATA_SECTION_NODE) {
+                            configuration.setConfigurationValue(configProperties.item(configPropertiesIndex)
+                                    .getNodeValue());
+                        }
+                    }
+                }
+                configuration.setConfigurationID(getConfigID(configurationNodes.item(configIndex)));
```
Comment (ayomawdb) [inline external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java]:
```suggestion
                configuration.setConfigurationID(getConfigId(configurationNodes.item(configIndex)));
```

File: external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java [hunk 7]
```diff
@@ -0,0 +1,601 @@
...
+     * @return
+     */
+    public static List<String> getExcludeProductFiles(Document childPropertiesDoc, String productPath) {
+        List<String> excludePathList = new ArrayList<String>();
+
+        NodeList excludePaths = childPropertiesDoc.getElementsByTagName(EXCLUDE_PATHS_TAG);
+
+        for (int excludePathsIndex = 0; excludePathsIndex < excludePaths.getLength(); excludePathsIndex++) {
+            NodeList excludePath = excludePaths.item(excludePathsIndex).getChildNodes();
+
```
Comment (ayomawdb) [inline external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java]:
```suggestion
```

File: external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java [hunk 8]
```diff
@@ -0,0 +1,601 @@
...
+            for (int excludePathIndex = 0; excludePathIndex < excludePath.getLength(); excludePathIndex++) {
+
```
Comment (ayomawdb) [inline external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java]:
```suggestion
```

File: external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java [hunk 9]
```diff
@@ -0,0 +1,601 @@
...
+     * @return a list of excluded configuration names
+     */
+    public static List<String> getExcludeConfigs(NodeList excludeConfigsNodeList)
+            throws ConfigCheckerException {
+        List<String> excludeConfigList = new ArrayList<>();
+        Node excludeConfigNode;
+
+        for (int excludeConfigNodeListIndex = 0; excludeConfigNodeListIndex <
+                excludeConfigsNodeList.getLength(); excludeConfigNodeListIndex++) {
+
```
Comment (ayomawdb) [inline external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java]:
```suggestion
```

File: external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java [hunk 10]
```diff
@@ -0,0 +1,601 @@
...
+            excludeConfigNode = excludeConfigsNodeList.item(excludeConfigNodeListIndex);
+
```
Comment (ayomawdb) [inline external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java]:
```suggestion
```

File: external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java [hunk 11]
```diff
@@ -0,0 +1,601 @@
...
+        for (int parentPropertiesListIndex = 0; parentPropertiesListIndex < parentPropertiesList.getLength();
+             parentPropertiesListIndex++) {
+
+            if (PARENT_PROPERTIES_TAG.equals(parentPropertiesList.item(parentPropertiesListIndex).
+                    getNodeName())) {
+                parentPropertyNodes = parentPropertiesList.item(parentPropertiesListIndex).getChildNodes();
+            }
+        }
+
```
Comment (ayomawdb) [inline external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java]:
```suggestion
```

File: external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java [hunk 12]
```diff
@@ -0,0 +1,601 @@
...
+        if (parentPropertyNodes == null) {
+            throw new ConfigCheckerException("Unable to find parent properties");
+        }
+
```
Comment (ayomawdb) [inline external/config-checker/src/main/java/org/wso2/security/tools/utils/Utils.java]:
```suggestion
```

**Comment by ayomawdb** [comment]
Please add param comments to all public methods and look at all locations with unnecessary new-lines. I have marked only some situations.

File: external/config-checker/Common/configurations/jaggery-conf-ref.xml [hunk 13]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (kasundharmadasa) [reply external/config-checker/Common/configurations/jaggery-conf-ref.xml]: Added the suggested improvements.

File: external/config-checker/Product/configurations/jaggery/jaggery-conf-ref.xml [hunk 14]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (kasundharmadasa) [reply external/config-checker/Product/configurations/jaggery/jaggery-conf-ref.xml]: Added the suggested improvements.

**Comment by ayomawdb** [comment]
> Also the main package used should be changed from `org.wso2.security.tools` to `org.wso2.security.tools.configchecker`.

@kasundharmadasa This comment is still valid.

**Comment by ayomawdb** [comment]
> Please add param comments to all public methods and look at all locations with unnecessary new-lines. I have marked only some situations.

@kasundharmadasa This comment is still valid. Let's add comments to util methods, interface methods and other public methods.

**Comment by ayomawdb** [comment]
Also please change the PR title to imperative form `Add configuration checker` would be more suitable.

File: internal/config-checker/config-checker-settings.yaml [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+#
+
+# Config Checker settings
+
+parentPropertiesFile : common/properties/parent.xml
+childPropertiesFile : product/properties/child.xml
+
+parsers:
+ xml : org.wso2.security.tools.parser.xml.XMLConfigurationParser
+ json : org.wso2.security.tools.parser.json.JSONConfigurationParser
```
Comment (ayomawdb) [inline internal/config-checker/config-checker-settings.yaml]:
```suggestion
 json : org.wso2.security.tools.configchecker.parser.json.JSONConfigurationParser
```

File: internal/config-checker/config-checker-settings.yaml [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+# limitations under the License.
+#
+
+# Config Checker settings
+
+parentPropertiesFile : common/properties/parent.xml
+childPropertiesFile : product/properties/child.xml
+
+parsers:
+ xml : org.wso2.security.tools.parser.xml.XMLConfigurationParser
```
Comment (ayomawdb) [inline internal/config-checker/config-checker-settings.yaml]:
```suggestion
 xml : org.wso2.security.tools.configchecker.parser.xml.XMLConfigurationParser
```

File: internal/config-checker/config-checker-settings.yaml [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+ json : org.wso2.security.tools.parser.json.JSONConfigurationParser
+
+reportGenerators:
+ text: org.wso2.security.tools.report.TextReportGenerator
```
Comment (ayomawdb) [inline internal/config-checker/config-checker-settings.yaml]:
```suggestion
 text: org.wso2.security.tools.configchecker.report.TextReportGenerator
```

**Pull request merged**